### PR TITLE
Change chainlocks enforcing/notification flow

### PR DESCRIPTION
**Pull request by UdjinM6**
Should enforce/notify when we have the block, not when we have only the header or trying to sign the tip (the later won't even work for non-masternodes).

Also, fix few related (dead)locks.

~NOTE: This includes #2759 atm, will rebase after it's merged. The actual changes are in 081a40c.~

### CODE REVIEW

File: src/llmq/quorums_chainlocks.cpp [hunk 1]
```diff
@@ -149,41 +151,36 @@ void CChainLocksHandler::ProcessNewChainLock(NodeId from, const llmq::CChainLock
...
         bestChainLockWithKnownBlock = bestChainLock;
         bestChainLockBlockIndex = pindex;
+        if (lastNotifyChainLockBlockIndex != bestChainLockBlockIndex) {
+            lastNotifyChainLockBlockIndex = bestChainLockBlockIndex;
+            pindexNotify = pindex;
+        }
     }
 
-    EnforceBestChainLock();
+    if (pindexNotify) {
```
Comment (PastaPastaPasta) [inline src/llmq/quorums_chainlocks.cpp]: I know we do this everywhere in the codebase, but I hate how we do this automatic integer/pointer to boolean conversions. 
Would prefer `if (pindexNotify != 0)`

Comment (UdjinM6) [reply src/llmq/quorums_chainlocks.cpp]: fixed :)

File: src/llmq/quorums_chainlocks.cpp [hunk 2]
```diff
@@ -337,31 +329,49 @@ void CChainLocksHandler::TrySignChainTip()
...
+    bool fEnforceAndNotify = false;
 
-    int64_t curTime = GetAdjustedTime();
+    {
+        LOCK(cs);
 
-    // We listen for NewPoWValidBlock so that we can collect all TX ids of all included TXs of newly received blocks
-    // We need this information later when we try to sign a new tip, so that we can determine if all included TXs are
-    // safe.
+        if (blockTxs.count(pindex->GetBlockHash())) {
```
Comment (PastaPastaPasta) [inline src/llmq/quorums_chainlocks.cpp]: Similar integer to boolean conversion.
`!= 0`

**Comment by UdjinM6** [comment]
Rebased. Changes without whitespaces https://github.com/dashpay/dash/pull/2762/files?w=1

File: src/llmq/quorums_chainlocks.cpp (hunk 2 above)
Comment (UdjinM6) [reply src/llmq/quorums_chainlocks.cpp]: Unlike in the previous case, this code wasn't touched (if you exclude whitespaces https://github.com/dashpay/dash/pull/2762/files?w=1) :P

**Comment by codablock** [comment]
Hmm, this will conflict quite a bit with what I'm working on right now (what I mentioned on Slack, the rework of how `EnforceBestChainLock` is called to make spork12 completely obsolete)

**Comment by UdjinM6** [comment]
@codablock I don't see why/how these two would conflict cause this one just fixes it in such a way that it would actually work on all nodes and for both `block->lock->notify` and `lock->block->notify` paths but maybe you came up with a better approach which also fixes these issues while working on spork12 obsoletion, I'm fine with that, le't see :)

fyi: rebased and force-pushed to trigger Travis build, no code changes.

**Comment by UdjinM6** [comment]
closing in fav of #2765

**Pull request closed without merging**